# -*- snakemake -*-
import os
from snakemakelib.config import update_sml_config, sml_rules_path, BaseConfig, get_sml_config

# Start by including the general bio.ngs snakefile
include: os.path.join(sml_rules_path(), 'bio/ngs', 'settings.rules')

sml_config = get_sml_config()

config_default = BaseConfig({
     'bio.ngs.methylseq.bismark' : BaseConfig({
        'ref' : sml_config['bio.ngs.settings']['db']['ref'],
        'cmd' : "bismark",
        'main' : BaseConfig({
            'options' : '--bam --bowtie2 --phred33-quals',
        }),
    }),
})

update_sml_config(config_default)

#### CFCMD bismark --bam --bowtie2 --phred33-quals
#### /apus/v1/a2010002/nobackup/biodata/genomes/Hsapiens/GRCh37/seq -1
#### 8_140207_AC3NVGACXX_P923_101_1_val_1.fq.gz -2
#### 8_140207_AC3NVGACXX_P923_101_2_val_2.fq.gz

sml_config = get_sml_config()

# Use subdictionary to enhance readability
cfg = sml_config['bio.ngs.methylseq.bismark']

rule bismark_main:
    """bismark: Run bismark methylation analysis"""
    params: options = cfg['main']['options'],
            cmd = cfg['cmd'],
            ref = cfg['ref']
    input: "{path}" + os.sep + "{prefix}" + sml_config['bio.ngs.settings']['read1_label'] + sml_config['bio.ngs.settings']['fastq_suffix'],\
    "{path}" + os.sep + "{prefix}" + sml_config['bio.ngs.settings']['read2_label'] + sml_config['bio.ngs.settings']['fastq_suffix']
    output: "{path}" + os.sep + "{prefix}.bismark"
    shell: "{params.cmd} {params.options} {params.ref} -1 {input[0]} -2 {input[1]} -o {wildcards.path}"
