
# -*- snakemake -*-
import os
from snakemakelib.utils import update_config, sml_rules_path

# Start by including the general snakefile
include: os.path.join(sml_rules_path(), 'base_settings.rules')

config_default = { 
	"gatk" : {
		"GATK_INDELREALIGNER_OPTIONS" : "-R $(GATK_REF)",
		"GATK_VARIANTFILTRATION_OPTIONS" : "",
		"GATK_REALIGN_TARGET_CREATOR_OPTIONS" : "-R $(GATK_REF)",
		"GATK_THREADS" : "$(THREADS)",
		"GATK_READBACKEDPHASING_OPTIONS" : "",
		"GATK_JAR" : "$(GATK_HOME)/GenomeAnalysisTK.jar",
		"GATK_COMMAND" : "java -Xmx$(GATK_JAVA_MEM) -Djava.io.tmpdir=$(GATK_JAVA_TMPDIR) -jar $(GATK_JAR)",
		"GATK_CLIPREADS_OPTIONS" : "",
		"GATK_UNIFIEDGENOTYPER_OPTIONS+" : "--dbsnp $(GATK_DBSNP)",
		"GATK_VARIANT_EVAL_OPTIONS" : "-ST Filter -l INFO --doNotUseAllStandardModules --evalModule CompOverlap --evalModule CountVariants --evalModule GenotypeConcordance --evalModule TiTvVariantEvaluator --evalModule ValidationReport --stratificationModule Filter",
		"GATK_TARGET_REGIONS" : "$(TARGET_REGIONS)",
		"GATK_BASERECALIBRATOR_OPTIONS" : "-R $(GATK_REF)",
		"GATK_KNOWN_SITES" : "$(GATK_DBSNP)",
		"GATK_SELECTSNPVARIANTS_OPTIONS" : "--selectTypeToInclude SNP",
		"GATK_UNIFIEDGENOTYPER_OPTIONS" : "-stand_call_conf 30.0 -stand_emit_conf 10.0  --downsample_to_coverage 30 --output_mode EMIT_VARIANTS_ONLY -glm BOTH -nt $(GATK_THREADS) -R $(GATK_REF)",
		"GATK_REF" : "$(REF)",
		"GATK_PRINTREADS_OPTIONS" : "-R $(GATK_REF)",
		"GATK_JAVA_MEM" : "$(JAVA_MEM)",
		"GATK_BAM_LIST:" : "",
		"GATK_VCFSUFFIX" : ".vcf",
		"GATK_HOME" : ".",
		"GATK_JAVA_TMPDIR" : "$(JAVA_TMPDIR)",
		"GATK_DBSNP" : "$(DBSNP)",
		},
	},
}

config = update_config(config, config_default)

rule rule_1:
	input: " {prefix}.bam"
	output: "{prefix}.vcf"
	shell: "$(GATK_COMMAND) -T UnifiedGenotyper $(GATK_UNIFIEDGENOTYPER_OPTIONS) -I $< -o $@.tmp && mv $@.tmp $@  && mv $@.tmp.idx $@.idx"
rule rule_2:
	input: " {prefix}.bam"
	output: "{prefix}.intervals"
	shell: "$(GATK_COMMAND) -T RealignerTargetCreator $(GATK_REALIGN_TARGET_CREATOR_OPTIONS) -I $< -o $@.tmp && mv $@.tmp $@"
rule rule_3:
	input: " {prefix}.bam {prefix}.intervals"
	output: "{prefix}.realign.bam"
	shell: "$(GATK_COMMAND) -T IndelRealigner $(GATK_INDELREALIGNER_OPTIONS) -o $@.tmp --targetIntervals $(word 2, $^) && mv $@.tmp $@  && mv $@.tmp.bai $@.bai"
rule rule_4:
	input: " {prefix}.bam {prefix}.bai"
	output: "{prefix}.recal_data.grp"
	shell: "$(GATK_COMMAND) -T BaseRecalibrator $(GATK_BASERECALIBRATOR_OPTIONS) $(KNOWN_SITES) -I $< -o $@.tmp && mv $@.tmp $@"
rule rule_5:
	input: " {prefix}.bam {prefix}.recal_data.grp"
	output: "{prefix}.recal.bam"
	shell: "$(GATK_COMMAND) -T PrintReads $(GATK_PRINTREADS_OPTIONS) -I $< -BQSR $(lastword $^) -o $@.tmp && mv $@.tmp $@ && mv $@.tmp.bai $@.bai"
rule rule_6:
	input: " {prefix}.bam {prefix}.bai"
	output: "{prefix}.clip.bam"
	shell: "$(GATK_COMMAND) -T ClipReads $(GATK_CLIPREADS_OPTIONS) -I $< -o $@.tmp && mv $@.tmp $@ && mv $@.tmp.bai $@.bai"
rule rule_7:
	input: " {prefix}.vcf"
	output: "{prefix}.filtered.vcf"
	shell: "$(GATK_COMMAND) -T VariantFiltration $(GATK_VARIANTFILTRATION_OPTIONS) -R $(GATK_REF) --variant $< --out $@.tmp && mv $@.tmp $@ && mv $@.tmp.idx $@.idx"
rule rule_8:
	input: " {prefix}.vcf"
	output: "{prefix}.eval_metrics"
	shell: "$(GATK_COMMAND) -T VariantEval $(GATK_VARIANT_EVAL_OPTIONS) -R $(GATK_REF) --eval $< -o $@.tmp && mv $@.tmp $@"
rule rule_9:
	input: " {prefix}.bam {prefix}.bai"
	output: "{prefix}.phased.vcf"
	shell: "$(GATK_COMMAND) -T ReadBackedPhasing $(GATK_READBACKEDPHASING_OPTIONS) -I $< --variant $*$(GATK_VCFSUFFIX) -R $(GATK_REF) -o $@.tmp && mv $@.tmp $@"
rule rule_10:
	input: " {prefix}.vcf"
	output: "{prefix}.snp.vcf"
	shell: "$(GATK_COMMAND) -T SelectVariants $(GATK_SELECTSNPVARIANTS_OPTIONS) --variant $< --out $@.tmp -R $(GATK_REF) && mv $@.tmp $@"
